File: nim_core_v2/ios/Classes/services/FLTUserService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import NIMSDK
import UIKit

enum V2UserAPITypeEnums: String {
  case getUserList
  case getUserListFromCloud
  case searchUserByOption
  case updateSelfUserProfile
  case addUserToBlockList
  case removeUserFromBlockList
  case getBlockList
  case checkBlock
}

let userClassName = "FLTUserService"
class FLTUserService: FLTBaseService, FLTService, V2NIMUserListener {
  override func onInitialized() {
    NIMSDK.shared().v2UserService.add(self)
  }

  deinit {
    NIMSDK.shared().v2UserService.remove(self)
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  func serviceName() -> String {
    ServiceType.UserService.rawValue
  }

  /**
   * 用户信息变更回调
   *
   * - Parameter users: 收到内容
   */
  func onUserProfileChanged(_ users: [V2NIMUser]) {
    notifyEvent(serviceName(), "onUserProfileChanged", ["userInfoList": users.map { $0.toDic() }])
  }

  /**
   * 黑名单添加通知
   *
   * - Parameter user: 加入黑名单用户
   */
  func onBlockListAdded(_ user: V2NIMUser) {
    let userInfo = user.toDic()
    print("onBlockListAdded -> \(userInfo)")
    notifyEvent(serviceName(), "onBlockListAdded", user.toDic())
  }

  /**
   * 黑名单移除通知
   * - Parameter accountId: 移除黑名的用户账号ID
   */
  func onBlockListRemoved(_ accountId: String) {
    notifyEvent(serviceName(), "onBlockListRemoved", ["userId": accountId])
  }

  /// 获取用户列表
  func getUserList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountIds = arguments["userIdList"] as? [String] else {
      parameterError(resultCallback)
      return
    }
    weak var weakSelf = self
    NIMSDK.shared().v2UserService.getUserList(accountIds) { users in
      var userList = [[String: Any]]()
      for user in users {
        userList.append(user.toDic())
      }
      weakSelf?.successCallBack(resultCallback, ["userInfoList": userList])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "getUserList error \(error.nserror.localizedDescription)")
    }
  }

  /// 从云端获取用户列表
  func getUserListFromCloud(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountIds = arguments["userIdList"] as? [String] else {
      parameterError(resultCallback)
      return
    }
    weak var weakSelf = self
    NIMSDK.shared().v2UserService.getUserList(fromCloud: accountIds, success: { users in
      var userList: [[String: Any]] = []
      for user in users {
        userList.append(user.toDic())
      }
      weakSelf?.successCallBack(resultCallback, ["userInfoList": userList])
    }, failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "getUserListFromCloud error \(error.nserror.localizedDescription)")
    })
  }

  /// 根据搜索条件搜索用户
  func searchUserByOption(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let userSearchOptionArguments = arguments["userSearchOption"] as? [String: Any] else {
      parameterError(resultCallback)
      return
    }
    weak var weakSelf = self
    let searchOption = V2NIMUserSearchOption.fromDic(userSearchOptionArguments)
    NIMSDK.shared().v2UserService.searchUser(by: searchOption) { users in
      var userList: [[String: Any]] = []
      for user in users {
        userList.append(user.toDic())
      }
      weakSelf?.successCallBack(resultCallback, ["userInfoList": userList])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "searchUserByOption error \(error.nserror.localizedDescription)")
    }
  }

  /// 更新自己的用户信息
  func updateSelfUserProfile(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let updateParamArguments = arguments["updateParam"] as? [String: Any] else {
      parameterError(resultCallback)
      return
    }
    let updateParams = V2NIMUserUpdateParams.fromDic(updateParamArguments)
    weak var weakSelf = self
    NIMSDK.shared().v2UserService.updateSelfUserProfile(updateParams) {
      weakSelf?.successCallBack(resultCallback, nil)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "updateSelfUserProfile error \(error.nserror.localizedDescription)")
    }
  }

  /// 添加用户到黑名单
  func addUserToBlockList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["userId"] as? String else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2UserService.addUser(toBlockList: accountId) {
      weakSelf?.successCallBack(resultCallback, nil)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "addUserToBlockList error \(error.nserror.localizedDescription)")
    }
  }

  /// 从黑名单移除用户
  func removeUserFromBlockList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["userId"] as? String else {
      parameterError(resultCallback)
      return
    }
    weak var weakSelf = self
    NIMSDK.shared().v2UserService.removeUser(fromBlockList: accountId) {
      weakSelf?.successCallBack(resultCallback, nil)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "removeUserFromBlockList error \(error.nserror.localizedDescription)")
    }
  }

  /// 获取黑名单列表
  func getBlockList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2UserService.getBlockList { blackList in
      weakSelf?.successCallBack(resultCallback, ["userIdList": blackList])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "getBlockList error \(error.nserror.localizedDescription)")
    }
  }

  /// 获取黑名单列表
  public func checkBlock(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountIds = arguments["accountIds"] as? [String] else {
      parameterError(resultCallback)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2UserService.checkBlock(accountIds) { result in
      weakSelf?.successCallBack(resultCallback, result)
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
      FLTALog.errorLog(userClassName, desc: "checkBlock error \(error.nserror.localizedDescription)")
    }
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback) {
    switch method {
    case V2UserAPITypeEnums.getUserList.rawValue:
      getUserList(arguments, resultCallback)
    case V2UserAPITypeEnums.getUserListFromCloud.rawValue:
      getUserListFromCloud(arguments, resultCallback)
    case V2UserAPITypeEnums.searchUserByOption.rawValue:
      searchUserByOption(arguments, resultCallback)
    case V2UserAPITypeEnums.updateSelfUserProfile.rawValue:
      updateSelfUserProfile(arguments, resultCallback)
    case V2UserAPITypeEnums.addUserToBlockList.rawValue:
      addUserToBlockList(arguments, resultCallback)
    case V2UserAPITypeEnums.removeUserFromBlockList.rawValue:
      removeUserFromBlockList(arguments, resultCallback)
    case V2UserAPITypeEnums.getBlockList.rawValue:
      getBlockList(arguments, resultCallback)
    case V2UserAPITypeEnums.checkBlock.rawValue:
      checkBlock(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }
}

File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTUserService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTConstant
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.toMap
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.v2.user.V2NIMUser
import com.netease.nimlib.sdk.v2.user.V2NIMUserListener
import com.netease.nimlib.sdk.v2.user.V2NIMUserService
import com.netease.nimlib.sdk.v2.user.option.V2NIMUserSearchOption
import com.netease.nimlib.sdk.v2.user.params.V2NIMUserUpdateParams.V2NIMUserUpdateParamsBuilder
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

class FLTUserService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {
    private val tag = "FLTUserService"
    override val serviceName = "UserService"

    /**
     * 根据用户账号列表获取用户资料 单次最大值150 只返回ID存在的用户，
     * 错误ID不返回 返回顺序以传入序为准（可以不做强制校验）
     * 先查询本地缓存，本地缺失或不足，再查询云端
     */
    private suspend fun getUserList(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val userIdList = arguments["userIdList"] as? List<String>
            if (userIdList == null) {
                cont.resume(
                    NimResult(code = FLTConstant.paramErrorCode, errorDetails = "getUserInfoList but the userIds is empty!")
                )
            } else {
                NIMClient.getService(V2NIMUserService::class.java).getUserList(
                    userIdList,
                    { result ->
                        cont.resume(
                            NimResult(
                                code = 0,
                                data = mutableMapOf("userInfoList" to result.map { it?.toMap() })
                            )
                        )
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "getUserInfoList failed!")
                        }
                    )
                }
            }
        }
    }

    /**
     * 根据用户账号列表从服务器获取用户资料 单次最大值150 只返回ID存在的用户，
     * 错误ID不返回 返回顺序以传入序为准（可以不做强制校验） 直接查询云端
     * 如果是协议错， 则整体返回错， 否则返回部分成功
     * 如果查询数据后，本地成员数据有更新， 则触发用户信息更新回调
     */
    private suspend fun getUserListFromCloud(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val userIdList = arguments["userIdList"] as? List<String>
            if (userIdList == null) {
                cont.resume(
                    NimResult(
                        code = FLTConstant.paramErrorCode,
                        errorDetails = "getUserListFromCloud but the userIds is empty!"
                    )
                )
            } else {
                NIMClient.getService(V2NIMUserService::class.java).getUserListFromCloud(
                    userIdList,
                    { result ->
                        cont.resume(
                            NimResult(
                                code = 0,
                                data =
                                mutableMapOf(
                                    "userInfoList" to result.map { it?.toMap() }
                                )
                            )
                        )
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "getUserListFromCloud failed!")
                        }
                    )
                }
            }
        }
    }

    /**
     * 更新自己的用户资料
     * 调用该Api后， SDK会抛出： onUserProfileChanged
     */
    private suspend fun updateSelfUserProfile(
        arguments: Map<String, *>
    ): NimResult<Nothing> {
        return suspendCancellableCoroutine { cont ->
            val updateParam = arguments["updateParam"] as Map<String, *>?
            if (updateParam == null) {
                cont.resume(NimResult(code = FLTConstant.paramErrorCode, errorDetails = "updateParam is null"))
            } else {
                val name = updateParam["name"] as? String
                val avatar = updateParam["avatar"] as? String
                val sign = updateParam["sign"] as? String
                val email = updateParam["email"] as? String
                val birth = updateParam["birthday"] as? String
                val mobile = updateParam["mobile"] as? String
                val gender = updateParam["gender"] as? Int
                val extension = updateParam["serverExtension"] as? String
                val param = V2NIMUserUpdateParamsBuilder.builder()
                if (name != null) {
                    param.withName(name)
                }
                if (avatar != null) {
                    param.withAvatar(avatar)
                }
                if (sign != null) {
                    param.withSign(sign)
                }
                if (email != null) {
                    param.withEmail(email)
                }
                if (birth != null) {
                    param.withBirthday(birth)
                }
                if (mobile != null) {
                    param.withMobile(mobile)
                }
                if (gender != null) {
                    param.withGender(gender)
                }
                if (extension != null) {
                    param.withServerExtension(extension)
                }

                NIMClient.getService(V2NIMUserService::class.java).updateSelfUserProfile(
                    param.build(),
                    {
                        cont.resume(NimResult.SUCCESS)
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "updateSelfUserProfile failed!")
                        }
                    )
                }
            }
        }
    }

    /**
     * 添加用户到黑名单中
     */
    private suspend fun addUserToBlockList(
        arguments: Map<String, *>
    ): NimResult<Nothing> {
        return suspendCancellableCoroutine { cont ->
            val userId = arguments["userId"] as? String
            if (userId == null) {
                cont.resume(
                    NimResult(
                        code = FLTConstant.paramErrorCode,
                        errorDetails = "addUserToBlockList but the userId is empty!"
                    )
                )
            } else {
                NIMClient.getService(V2NIMUserService::class.java).addUserToBlockList(
                    userId,
                    {
                        cont.resume(NimResult.SUCCESS)
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "addUserToBlockList failed!")
                        }
                    )
                }
            }
        }
    }

    /**
     * 从黑名单中移除用户
     */
    private suspend fun removeUserFromBlockList(
        arguments: Map<String, *>
    ): NimResult<Nothing> {
        return suspendCancellableCoroutine { cont ->
            val userId = arguments["userId"] as? String
            if (userId == null) {
                cont.resume(
                    NimResult(
                        code = FLTConstant.paramErrorCode,
                        errorDetails = "removeUserFromBlockList but the userId is empty!"
                    )
                )
            } else {
                NIMClient.getService(V2NIMUserService::class.java).removeUserFromBlockList(
                    userId,
                    {
                        cont.resume(NimResult.SUCCESS)
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "removeUserFromBlockList failed!")
                        }
                    )
                }
            }
        }
    }

    /**
     * 获取黑名单列表
     */
    private suspend fun getBlockList(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMUserService::class.java).getBlockList(
                { result ->
                    cont.resume(NimResult(code = 0, data = mutableMapOf("userIdList" to result)))
                }
            ) { error ->
                cont.resume(
                    if (error != null) {
                        NimResult(code = error.code, errorDetails = error.desc)
                    } else {
                        NimResult(code = -1, errorDetails = "getBlockList failed!")
                    }
                )
            }
        }
    }

    /**
     * 根据关键字搜索用户信息
     * Params:
     * userSearchOption – 用户搜索相关参数
     */
    private suspend fun searchUserByOption(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val userSearchOption = arguments["userSearchOption"] as? Map<String, *>
            val keyword = userSearchOption?.get("keyword") as? String
            if (keyword == null) {
                cont.resume(
                    NimResult(
                        code = FLTConstant.paramErrorCode,
                        errorDetails = "searchUserByOption but the keyword is empty!"
                    )
                )
            } else {
                val searchName = userSearchOption?.get("searchName") as? Boolean
                val searchAccountId = userSearchOption?.get("searchAccountId") as? Boolean
                val searchMobile = userSearchOption?.get("searchMobile") as? Boolean
                val searchOption =
                    V2NIMUserSearchOption.V2NIMUserSearchOptionBuilder.builder(keyword)
                if (searchName != null) {
                    searchOption.withSearchName(searchName)
                }
                if (searchAccountId != null) {
                    searchOption.withSearchAccountId(searchAccountId)
                }
                if (searchMobile != null) {
                    searchOption.withSearchMobile(searchMobile)
                }
                NIMClient.getService(V2NIMUserService::class.java).searchUserByOption(
                    searchOption.build(),
                    { result ->
                        cont.resume(
                            NimResult(
                                code = 0,
                                data =
                                mutableMapOf(
                                    "userInfoList" to result.map { it?.toMap() }
                                )
                            )
                        )
                    }
                ) { error ->
                    cont.resume(
                        if (error != null) {
                            NimResult(code = error.code, errorDetails = error.desc)
                        } else {
                            NimResult(code = -1, errorDetails = "searchUserByOption failed!")
                        }
                    )
                }
            }
        }
    }

    init {
        nimCore.onInitialized {
            NIMClient.getService(V2NIMUserService::class.java).addUserListener(userListener)
        }
        registerFlutterMethodCalls(
            "getUserList" to ::getUserList,
            "getUserListFromCloud" to ::getUserListFromCloud,
            "updateSelfUserProfile" to ::updateSelfUserProfile,
            "addUserToBlockList" to ::addUserToBlockList,
            "removeUserFromBlockList" to ::removeUserFromBlockList,
            "getBlockList" to ::getBlockList,
            "searchUserByOption" to ::searchUserByOption
        )
    }

    private val userListener =
        object : V2NIMUserListener {
            override fun onUserProfileChanged(users: MutableList<V2NIMUser>?) {
                notifyEvent(
                    "onUserProfileChanged",
                    mutableMapOf(
                        "userInfoList" to
                            users?.map { it.toMap() }
                                ?.toList()
                    )
                )
            }

            override fun onBlockListAdded(user: V2NIMUser?) {
                if (user != null) {
                    notifyEvent("onBlockListAdded", user!!.toMap())
                } else {
                    notifyEvent("onBlockListAdded", mutableMapOf())
                }
            }

            override fun onBlockListRemoved(accountId: String?) {
                notifyEvent(
                    "onBlockListRemoved",
                    mutableMapOf(
                        "userId" to accountId
                    )
                )
            }
        }
}
